File: mobile/lib/modules/today/models/usage.dart
/// Mirrors the backend `CurrentWeekUsage` (`GET /usage/current-week`) — the
/// weekly free-tier counters that drive the 75% / 90% / 100% usage banner.
library;

class UsageCounters {
  const UsageCounters({
    required this.used,
    required this.limit,
    required this.remaining,
    required this.percentage,
  });

  final int used;
  final int limit;
  final int remaining;
  final double percentage; // 0–100

  factory UsageCounters.fromJson(Map<String, dynamic> json) => UsageCounters(
        used: json['used'] as int,
        limit: json['limit'] as int,
        remaining: json['remaining'] as int,
        percentage: (json['percentage'] as num).toDouble(),
      );
}

class CurrentWeekUsage {
  const CurrentWeekUsage({
    required this.outfits,
    required this.mixAndMatch,
    required this.nextReset,
    required this.subscriptionTier,
    this.lastReset,
  });

  final UsageCounters outfits;
  final UsageCounters mixAndMatch;
  final DateTime nextReset;
  final String subscriptionTier; // free | pro
  final DateTime? lastReset;

  bool get isPro => subscriptionTier == 'pro';

  factory CurrentWeekUsage.fromJson(Map<String, dynamic> json) {
    return CurrentWeekUsage(
      outfits: UsageCounters.fromJson(json['outfits'] as Map<String, dynamic>),
      mixAndMatch:
          UsageCounters.fromJson(json['mix_and_match'] as Map<String, dynamic>),
      nextReset: DateTime.parse(json['next_reset'] as String),
      subscriptionTier: json['subscription_tier'] as String? ?? 'free',
      lastReset: json['last_reset'] == null
          ? null
          : DateTime.parse(json['last_reset'] as String),
    );
  }
}

File: backend/app/schemas/usage.py
"""Phase 6d — usage tracking request/response shapes.

`UsageResource` keys the limit family being checked. `LimitReachedDetail` is
the body shape served back as a 429; the client uses `resets_at` to render
the countdown timer in the "100% Usage Limit Reached" screen.
"""
from __future__ import annotations

from datetime import date, datetime
from typing import Literal

from pydantic import BaseModel

UsageResource = Literal["outfits", "mix_and_match", "buy_dont_buy", "advisor"]
SubscriptionTier = Literal["free", "pro"]


class UsageCounters(BaseModel):
    """Snapshot of one resource within the current week window."""

    used: int
    limit: int
    remaining: int
    percentage: float


class CurrentWeekUsage(BaseModel):
    week_start_date: date
    outfits: UsageCounters
    mix_and_match: UsageCounters
    buy_dont_buy: UsageCounters
    advisor: UsageCounters
    last_reset: datetime | None
    next_reset: datetime
    subscription_tier: SubscriptionTier


class LimitReachedDetail(BaseModel):
    error: Literal["limit_reached"] = "limit_reached"
    resource: UsageResource
    used: int
    limit: int
    resets_at: datetime
    message: str


class ProRequiredDetail(BaseModel):
    error: Literal["pro_required"] = "pro_required"
    feature: str
    message: str
